src/arith/transitive_comparison_analyzer.cc

```diff
@@ -0,0 +1,683 @@
+    case CompareResult::kNE:
+      return CompareResult::kNE;
+    case CompareResult::kUnknown:
+      return CompareResult::kUnknown;
+    default:
+      LOG(FATAL) << "Invalid CompareResult: " << static_cast<int>(res);
+      return CompareResult::kInconsistent;
+  }
+}
+
+// Internal utility, return the CompareResult resulting from negating
+// the comparison.
+CompareResult Negate(CompareResult res) {
+  switch (res) {
+    case CompareResult::kInconsistent:
+      return CompareResult::kInconsistent;
+    case CompareResult::kUnknown:
+      return CompareResult::kUnknown;
+    default:
+      return CompareResult(~static_cast<int>(res) & static_cast<int>(CompareResult::kUnknown));
+  }
+}
+
+// Internal utility, extract constant offsets out of the two sides of
+// a comparison.  Given lhs and rhs, return a tuple of three elements
+// (lhs_inner, rhs_inner, offset), such that (lhs OP rhs) and
+// (lhs_inner OP rhs_inner + offset) are equivalent.
+std::tuple<PrimExpr, PrimExpr, int64_t> ExtractOffsets(const PrimExpr& lhs, const PrimExpr& rhs) {
+  auto extract_offset = [](const PrimExpr& expr) -> std::pair<PrimExpr, int64_t> {
+    PVar<PrimExpr> x;
+    PVar<IntImm> c;
+    if ((x + c).Match(expr)) {
+      return {x.Eval(), c.Eval()->value};
+    } else if ((x - c).Match(expr)) {
+      return {x.Eval(), -c.Eval()->value};
+    } else if (c.Match(expr)) {
+      return {0, c.Eval()->value};
+    } else {
+      return {expr, 0};
+    }
+  };
+
+  auto lhs_split = extract_offset(lhs);
+  auto rhs_split = extract_offset(rhs);
+  return {lhs_split.first, rhs_split.first, rhs_split.second - lhs_split.second};
+}
+
+}  // namespace
+
+std::optional<TransitiveComparisonAnalyzer::Impl::Comparison>
+TransitiveComparisonAnalyzer::Impl::FromExpr(const PrimExpr& expr) {
+  CompareResult res;
+  PVar<PrimExpr> x, y;
+  if ((x <= y).Match(expr)) {
+    res = CompareResult::kLE;
+  } else if ((x >= y).Match(expr)) {
+    res = CompareResult::kGE;
+  } else if ((x < y).Match(expr)) {
+    res = CompareResult::kLT;
+  } else if ((x > y).Match(expr)) {
+    res = CompareResult::kGT;
+  } else if ((x == y).Match(expr)) {
+    res = CompareResult::kEQ;
+  } else if ((x != y).Match(expr)) {
+    res = CompareResult::kNE;
+  } else {
+    return std::nullopt;
+  }
+
+  PrimExpr lhs_expr = x.Eval();
+  PrimExpr rhs_expr = y.Eval();
+
+  if (lhs_expr.as<IntImmNode>() && rhs_expr.as<IntImmNode>()) {
+    return std::nullopt;
+  }
+
+  auto [lhs, rhs, offset] = ExtractOffsets(lhs_expr, rhs_expr);
+  Key lhs_key = ExprToKey(lhs);
+  Key rhs_key = ExprToKey(rhs);
+
+  return Comparison(lhs_key, rhs_key, offset, res);
+}
+
+TransitiveComparisonAnalyzer::Impl::Comparison::Comparison(Key lhs, Key rhs, int64_t offset,
+                                                           CompareResult result)
+    : lhs_(lhs), rhs_(rhs), offset_(offset), result_(result) {
+  if (result_ == CompareResult::kLT) {
+    result_ = CompareResult::kLE;
+    offset_ -= 1;
```

Comment:
Sounds good, and added a description in the constructor, along with an explanation of why the internal representation normalizes everything to GE/LE, instead of normalizing to GT/LT.